clippy_lints/src/manual_option_folding.rs

```diff
@@ -0,0 +1,98 @@
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::ty::is_type_diagnostic_item;
+use rustc_hir as hir;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_span::{sym, Symbol};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for cases where `unwrap_or_else` can be used in lieu of
+    /// `get_or_insert_with` followed by `unwrap`/`unwrap_unchecked`/`expect`.
+    ///
+    /// ### Why is this bad?
+    /// It is more concise to use `unwrap_or_else`, and using `unwrap_or_else`
+    /// instead of `unwrap_unchecked` eliminates the need for an `unsafe`
+    /// block.
+    ///
+    /// ### Example
+    /// ```rust
+    /// let mut opt: Option<i32> = None;
+    /// opt.get_or_insert(42);
+    /// let res = unsafe { opt.unwrap_unchecked() };
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let opt: Option<i32> = None;
+    /// let res: i32 = opt.unwrap_or(42);
+    /// ```
+    #[clippy::version = "1.74.0"]
+    pub MANUAL_OPTION_FOLDING,
+    style,
+    "manual implementation of `Option::unwrap_or_else`"
+}
+
+impl_lint_pass!(ManualOptionFolding<'_> => [MANUAL_OPTION_FOLDING]);
+
+pub struct ManualOptionFolding<'tcx> {
+    get_call: Option<&'tcx hir::Expr<'tcx>>,
+    recv: Option<&'tcx hir::Expr<'tcx>>,
+    get_method_name: Option<Symbol>,
+}
+
+impl<'tcx> ManualOptionFolding<'tcx> {
+    pub fn new() -> Self {
+        Self {
+            get_call: None,
+            recv: None,
+            get_method_name: None,
+        }
+    }
+}
+
+impl<'tcx> LateLintPass<'tcx> for ManualOptionFolding<'tcx> {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx hir::Expr<'tcx>) {
+        if !expr.span.from_expansion()
+            && let hir::ExprKind::MethodCall(path, recv, ..) = expr.kind
+            && let ty = cx.typeck_results().expr_ty(recv).peel_refs()
+            && is_type_diagnostic_item(cx, ty, sym::Option)
+        {
+            if path.ident.name == sym!(get_or_insert)
+                || path.ident.name == sym!(get_or_insert_with)
+                || path.ident.name == sym!(get_or_insert_default)
+            {
+                self.get_call = Some(expr);
+                self.recv = Some(recv);
+                self.get_method_name = Some(path.ident.name);
+            } else if let Some(get_call) = self.get_call.take()
```

Comment:
This `take` is only executed on method expressions that fulfill the prerequisites at the beginning of this function.

This means that any number of expressions could between this and the previous `insert`. This can sadly lead to false positives, as in this example:

```rs
fn magic(condition: bool) -> u32 {
    let mut option = None;
    if condition {
        option.insert(19);
    }
    option.unwrap()
}
```

I think the best solution would be to check for block expressions and then iterate over the statements using a window of 2. That would only catch consequitve `option.insert` & `option.unwrap` things, but anything else would be way harder to implement and could result in false positives. Maybe you can take a look at [`clippy::vec_init_then_push`] here: [`clippy_lints/src/vec_init_then_push.rs`](https://github.com/rust-lang/rust-clippy/blob/master/clippy_lints/src/vec_init_then_push.rs#L42). It's definitly a more complicated implementation, but that would be the best way AFAIK.